File: apps/backend/routers/shopping.py
"""
购物清单路由 - PostgreSQL版
"""
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional

router = APIRouter()

class ShoppingItemCreate(BaseModel):
    name: str
    quantity: str = "1"
    unit: str = "件"
    category: str = "general"
    priority: str = "normal"
    notes: str = ""
    current_stock: float = 0
    target_stock: float = 0
    restock_threshold: float = 0
    is_favorite: bool = False
    family_id: str = ""


def get_agent():
    from backend.main import get_agent as _get_agent
    return _get_agent()


@router.get("/shopping")
async def get_shopping_list(family_id: str = ""):
    agent = get_agent()
    items = agent.shopping_list.get_items(family_id=family_id)
    # 添加 purchased 布尔字段（前端兼容）
    for item in items:
        item['purchased'] = item.get('status') == 'purchased'
    return {"items": items}


@router.post("/shopping", status_code=201)
async def add_shopping_item(item_data: ShoppingItemCreate):
    agent = get_agent()
    agent.shopping_list.add_item(
        name=item_data.name,
        quantity=item_data.quantity,
        unit=item_data.unit,
        category=item_data.category,
        priority=item_data.priority,
        notes=item_data.notes,
        current_stock=item_data.current_stock,
        target_stock=item_data.target_stock,
        restock_threshold=item_data.restock_threshold,
        is_favorite=item_data.is_favorite,
        family_id=getattr(item_data, "family_id", "")
    )
    return {"success": True, "message": f"已添加: {item_data.name}"}


@router.put("/shopping/{item_id}")
async def update_shopping_item(item_id: int, item_data: ShoppingItemCreate, family_id: str = ""):
    agent = get_agent()
    agent.shopping_list.update_item(item_id, family_id=family_id, **item_data.model_dump(exclude={"family_id"}))
    return {"success": True, "message": f"已更新: {item_data.name}"}


@router.post("/shopping/{item_id}/toggle")
async def toggle_shopping_item(item_id: int, family_id: str = ""):
    agent = get_agent()
    agent.shopping_list.toggle_purchased(item_id, family_id=family_id)
    return {"success": True}


@router.delete("/shopping/{item_name}")
async def remove_shopping_item(item_name: str, family_id: str = ""):
    agent = get_agent()
    items = agent.shopping_list.get_items(family_id=family_id)
    for item in items:
        if item.get('name') == item_name:
            agent.shopping_list.remove_item(item['id'], family_id=family_id)
            return {"success": True, "message": f"已删除: {item_name}"}
    return {"success": False, "message": f"未找到: {item_name}"}


@router.get("/shopping/stats")
async def get_shopping_stats(family_id: str = ""):
    agent = get_agent()
    stats = agent.shopping_list.get_shopping_summary(family_id=family_id)
    return stats


@router.get("/suggestions/{member_name}")
async def get_smart_suggestions(member_name: str):
    agent = get_agent()
    suggestions = agent.shopping_list.get_smart_suggestions(member_name)
    return {"suggestions": suggestions}

File: apps/miniapp/pages/shopping/shopping.js
const { request } = require('../../utils/request')

const CATEGORY_LABELS = {
  food: '食品食材',
  daily: '日用品',
  health: '医药保健',
  home: '家居用品',
  electronics: '电子产品',
  clothing: '服装鞋帽',
  other: '其他',
  general: '其他',
}

Page({
  data: {
    loading: true,
    submitting: false,
    items: [],
    pendingItems: [],
    purchasedItems: [],
    restockItems: [],
    familyId: '',
    memberName: '',
    itemName: '',
    itemQuantity: '1',
    stats: {
      total_items: 0,
      unpurchased: 0,
      purchased: 0,
      restock_needed: 0,
    },
  },

  onLoad() {
    wx.setNavigationBarTitle({ title: '购物清单' })
    const memberName = wx.getStorageSync('member_name') || ''
    const familyId = wx.getStorageSync('family_id') || ''
    if (!memberName || !familyId) {
      wx.redirectTo({ url: '/pages/auth/auth' })
      return
    }
    this.setData({ memberName, familyId })
    this.refresh()
  },

  onShow() {
    if (this.data.familyId) {
      this.refresh()
    }
  },

  setField(e) {
    const { field } = e.currentTarget.dataset
    this.setData({ [field]: e.detail.value })
  },

  refresh() {
    this.setData({ loading: true })
    Promise.all([this.loadItems(), this.loadStats()])
      .then(() => {})
      .catch(() => {})
      .then(() => this.setData({ loading: false }))
  },

  loadItems() {
    return request({
      path: '/api/shopping',
      method: 'GET',
      data: { family_id: this.data.familyId },
    }).then((data) => {
      const items = (data.items || []).map((item) => {
        const currentStock = Number(item.current_stock || 0)
        const threshold = Number(item.restock_threshold || 0)
        const purchased = item.purchased || item.status === 'purchased'
        return {
          ...item,
          purchased,
          category_text: CATEGORY_LABELS[item.category] || item.category || '其他',
          quantity_text: `${item.quantity || '1'} ${item.unit || '件'}`,
          priority_text: item.priority === 'high' ? '重要' : item.priority === 'low' ? '低优先' : '普通',
          added_by_text: item.added_by || '未记录',
          is_restock: !purchased && currentStock <= threshold && threshold > 0,
        }
      })
      this.setData({
        items,
        pendingItems: items.filter((item) => !item.purchased),
        purchasedItems: items.filter((item) => item.purchased),
        restockItems: items.filter((item) => item.is_restock),
      })
    })
  },

  loadStats() {
    return request({
      path: '/api/shopping/stats',
      method: 'GET',
      data: { family_id: this.data.familyId },
    }).then((data) => {
      this.setData({
        stats: {
          total_items: data.total_items || 0,
          unpurchased: data.unpurchased || 0,
          purchased: data.purchased || 0,
          restock_needed: data.restock_needed || 0,
        },
      })
    })
  },

  addItem() {
    const name = this.data.itemName.trim()
    if (!name) {
      wx.showToast({ title: '请输入物品名称', icon: 'none' })
      return
    }
    this.setData({ submitting: true })
    request({
      path: '/api/shopping',
      method: 'POST',
      data: {
        name,
        quantity: this.data.itemQuantity || '1',
        unit: '件',
        category: 'daily',
        priority: 'normal',
        notes: '',
        added_by: this.data.memberName,
        family_id: this.data.familyId,
      },
    })
      .then((data) => {
        if (data && data.success) {
          wx.showToast({ title: '已加入清单', icon: 'success' })
          this.setData({ itemName: '', itemQuantity: '1' })
          this.refresh()
          return
        }
        wx.showToast({ title: (data && data.message) || '添加失败', icon: 'none' })
      })
      .catch(() => {})
      .then(() => this.setData({ submitting: false }))
  },

  toggleItem(e) {
    const itemId = e.currentTarget.dataset.itemid
    if (!itemId) return
    request({
      path: `/api/shopping/${encodeURIComponent(itemId)}/toggle?family_id=${encodeURIComponent(this.data.familyId)}`,
      method: 'POST',
    })
      .then((data) => {
        if (data && data.success) {
          this.refresh()
          return
        }
        wx.showToast({ title: '操作失败', icon: 'none' })
      })
      .catch(() => {})
  },

  openChatAdd() {
    wx.navigateTo({
      url: `/pages/webview/webview?route=${encodeURIComponent('/chat?prompt=/shopping ')}&title=${encodeURIComponent('添加采购')}`,
    })
  },
})
